sweepbatcher/sweep_batch.go

```diff
@@ -561,46 +520,151 @@ func (b *batch) addSweep(ctx context.Context, sweep *sweep) (bool, error) {
 			int32(math.Abs(float64(sweep.timeout - s.timeout)))
 
 		if timeoutDistance > b.cfg.maxTimeoutDistance {
-			b.Infof("too long timeout distance between the "+
-				"batch and sweep %x: %d > %d",
+			return false, fmt.Errorf("too long timeout distance "+
+				"between the batch and sweep %x: %d > %d",
 				sweep.swapHash[:6], timeoutDistance,
 				b.cfg.maxTimeoutDistance)
+		}
+	}
+
+	// Everything is ok, the sweep can be added to the batch.
+	return false, nil
+}
+
+// addSweeps tries to add sweeps to the batch. If this is the first sweep being
+// added to the batch then it also sets the primary sweep ID. It returns if the
+// sweeps were accepted to the batch.
+func (b *batch) addSweeps(ctx context.Context, sweeps []*sweep) (bool, error) {
+	done, err := b.scheduleNextCall()
+	defer done()
+	if err != nil {
+		return false, err
+	}
+
+	// Track if there is an existing sweep and a new sweep among the sweeps.
+	var numExisting, numNew int
+	for _, s := range sweeps {
+		existing, err := b.checkSweepToAdd(ctx, s)
+		if err != nil {
+			b.Infof("Failed to add sweep %v to batch %d: %v",
+				s.outpoint, b.id, err)
+
+			return false, nil
+		}
+		if existing {
+			numExisting++
+		} else {
+			numNew++
+		}
+	}
+
+	// Make sure the whole group is either new or existing. If this is not
+	// the case, this might be a bug, so print a warning.
+	if numExisting > 0 && numNew > 0 {
+		b.Warnf("There are %d existing and %d new sweeps among the "+
+			"group. They must not be mixed.", numExisting, numNew)
+
+		return false, nil
+	}
+
+	// Make sure all the sweeps spend different outpoints.
+	outpointsSet := make(map[wire.OutPoint]struct{}, len(sweeps))
+	for _, s := range sweeps {
+		if _, has := outpointsSet[s.outpoint]; has {
+			b.Warnf("Multiple sweeps spend outpoint %v", s.outpoint)
 
 			return false, nil
 		}
+		outpointsSet[s.outpoint] = struct{}{}
 	}
 
 	// Past this point we know that a new incoming sweep passes the
 	// acceptance criteria and is now ready to be added to this batch.
 
-	// If this is the first sweep being added to the batch, make it the
-	// primary sweep.
-	if b.primarySweepID == zeroSweepID {
-		b.primarySweepID = sweep.outpoint
-		b.cfg.batchConfTarget = sweep.confTarget
-		b.rbfCache.FeeRate = sweep.minFeeRate
-		b.rbfCache.SkipNextBump = true
-
-		// We also need to start the spend monitor for this new primary
-		// sweep.
-		err := b.monitorSpend(ctx, *sweep)
-		if err != nil {
-			return false, err
+	// For an existing group, update the sweeps in the batch.
+	if numExisting > 0 {
```

Comment:
nit: ... and `countExisting == len(sweeps)`

Reply:
Fixed!